File: packages/performance/ios/Plugin/FirebasePerformancePlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(FirebasePerformancePlugin)
public class FirebasePerformancePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "FirebasePerformancePlugin"
    public let jsName = "FirebasePerformance"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "startTrace", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopTrace", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "incrementMetric", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setEnabled", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "isEnabled", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "putAttribute", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getAttribute", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getAttributes", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removeAttribute", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "putMetric", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getMetric", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "record", returnType: CAPPluginReturnPromise)
    ]
    public let errorTraceNameMissing = "traceName must be provided."
    public let errorMetricNameMissing = "metricName must be provided."
    public let errorEnabledMissing = "enabled must be provided."
    public let errorTraceNameAlreadyAssigned = "traceName already assigned."
    public let errorTraceNotFound = "No trace was found with the provided traceName."
    public let errorAttributeMissing = "attribute must be provided."
    public let errorValueMissing = "value must be provided."
    public let errorNumMissing = "num must be provided."
    public let errorStartTimeMissing = "startTime must be provided."
    public let errorDurationMissing = "duration must be provided."
    private var implementation: FirebasePerformance?

    override public func load() {
        implementation = FirebasePerformance()
    }

    @objc func startTrace(_ call: CAPPluginCall) {
        guard let traceName = call.getString("traceName") else {
            call.reject(errorTraceNameMissing)
            return
        }
        let trace = implementation?.getTraceByName(traceName)
        guard trace == nil else {
            call.reject(errorTraceNameAlreadyAssigned)
            return
        }
        implementation?.startTrace(traceName)
        call.resolve()
    }

    @objc func stopTrace(_ call: CAPPluginCall) {
        guard let traceName = call.getString("traceName") else {
            call.reject(errorTraceNameMissing)
            return
        }
        let trace = implementation?.getTraceByName(traceName)
        guard trace != nil else {
            call.reject(errorTraceNotFound)
            return
        }
        implementation?.stopTrace(traceName)
        call.resolve()
    }

    @objc func incrementMetric(_ call: CAPPluginCall) {
        guard let traceName = call.getString("traceName") else {
            call.reject(errorTraceNameMissing)
            return
        }
        guard let metricName = call.getString("metricName") else {
            call.reject(errorMetricNameMissing)
            return
        }
        let trace = implementation?.getTraceByName(traceName)
        guard trace != nil else {
            call.reject(errorTraceNotFound)
            return
        }
        let incrementBy = call.getInt("incrementBy") ?? 1
        implementation?.incrementMetric(traceName, metricName, incrementBy)
        call.resolve()
    }

    @objc func setEnabled(_ call: CAPPluginCall) {
        guard let enabled = call.getBool("enabled") else {
            call.reject(errorEnabledMissing)
            return
        }
        implementation?.setEnabled(enabled)
        call.resolve()
    }

    @objc func isEnabled(_ call: CAPPluginCall) {
        let enabled = implementation?.isEnabled()
        var result = JSObject()
        result["enabled"] = enabled
        call.resolve(result)
    }

    @objc func putAttribute(_ call: CAPPluginCall) {
        guard let traceName = call.getString("traceName") else {
            call.reject(errorTraceNameMissing)
            return
        }
        guard let attribute = call.getString("attribute") else {
            call.reject(errorAttributeMissing)
            return
        }
        guard let value = call.getString("value") else {
            call.reject(errorValueMissing)
            return
        }
        let trace = implementation?.getTraceByName(traceName)
        guard trace != nil else {
            call.reject(errorTraceNotFound)
            return
        }
        FirebasePerformance.putAttribute(trace!, attribute, value)
        call.resolve()
    }

    @objc func getAttribute(_ call: CAPPluginCall) {
        guard let traceName = call.getString("traceName") else {
            call.reject(errorTraceNameMissing)
            return
        }
        guard let attribute = call.getString("attribute") else {
            call.reject(errorAttributeMissing)
            return
        }
        let trace = implementation?.getTraceByName(traceName)
        guard trace != nil else {
            call.reject(errorTraceNotFound)
            return
        }
        var result = JSObject()
        result["value"] = FirebasePerformance.getAttribute(trace!, attribute) ?? NSNull()
        call.resolve(result)
    }

    @objc func getAttributes(_ call: CAPPluginCall) {
        guard let traceName = call.getString("traceName") else {
            call.reject(errorTraceNameMissing)
            return
        }
        let trace = implementation?.getTraceByName(traceName)
        guard trace != nil else {
            call.reject(errorTraceNotFound)
            return
        }
        call.resolve(["result": FirebasePerformance.getAttributes(trace!)])
    }

    @objc func removeAttribute(_ call: CAPPluginCall) {
        guard let traceName = call.getString("traceName") else {
            call.reject(errorTraceNameMissing)
            return
        }
        guard let attribute = call.getString("attribute") else {
            call.reject(errorAttributeMissing)
            return
        }
        let trace = implementation?.getTraceByName(traceName)
        guard trace != nil else {
            call.reject(errorTraceNotFound)
            return
        }
        FirebasePerformance.removeAttribute(trace!, attribute)
        call.resolve()
    }

    @objc func putMetric(_ call: CAPPluginCall) {
        guard let traceName = call.getString("traceName") else {
            call.reject(errorTraceNameMissing)
            return
        }
        guard let metricName = call.getString("metricName") else {
            call.reject(errorMetricNameMissing)
            return
        }
        guard let num = call.getDouble("num") else {
            call.reject(errorNumMissing)
            return
        }
        let trace = implementation?.getTraceByName(traceName)
        guard trace != nil else {
            call.reject(errorTraceNotFound)
            return
        }
        FirebasePerformance.putMetric(trace!, metricName, num)
        call.resolve()
    }

    @objc func getMetric(_ call: CAPPluginCall) {
        guard let traceName = call.getString("traceName") else {
            call.reject(errorTraceNameMissing)
            return
        }
        guard let metricName = call.getString("metricName") else {
            call.reject(errorMetricNameMissing)
            return
        }
        let trace = implementation?.getTraceByName(traceName)
        guard trace != nil else {
            call.reject(errorTraceNotFound)
            return
        }
        let value = FirebasePerformance.getMetric(trace!, metricName)
        call.resolve(["value": value])
    }

    @objc func record(_ call: CAPPluginCall) {
        call.unimplemented("Not implemented on iOS.")
    }
}

File: packages/performance/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/performance/FirebasePerformancePlugin.java
package io.capawesome.capacitorjs.plugins.firebase.performance;

import com.getcapacitor.JSObject;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.google.firebase.perf.metrics.Trace;
import java.util.Map;

@CapacitorPlugin(name = "FirebasePerformance")
public class FirebasePerformancePlugin extends Plugin {

    public static final String TAG = "FirebasePerformance";
    public static final String ERROR_TRACE_NAME_MISSING = "traceName must be provided.";
    public static final String ERROR_METRIC_NAME_MISSING = "metricName must be provided.";
    public static final String ERROR_TRACE_NAME_ALREADY_ASSIGNED = "traceName already assigned.";
    public static final String ERROR_ENABLED_MISSING = "enabled must be provided.";
    public static final String ERROR_TRACE_NOT_FOUND = "No trace was found with the provided traceName.";
    public static final String ERROR_ATTRIBUTE_MISSING = "attribute must be provided.";
    public static final String ERROR_VALUE_MISSING = "value must be provided.";
    public static final String ERROR_NUM_MISSING = "num must be provided.";
    public static final String ERROR_START_TIME_MISSING = "startTime must be provided.";
    public static final String ERROR_DURATION_MISSING = "duration must be provided.";
    public static final String ERROR_INVALID_METRIC_VALUE = "provided metric value is not a number.";
    private FirebasePerformance implementation = new FirebasePerformance();

    @PluginMethod
    public void startTrace(PluginCall call) {
        try {
            String traceName = call.getString("traceName");
            if (traceName == null) {
                call.reject(ERROR_TRACE_NAME_MISSING);
                return;
            }
            Trace trace = implementation.getTraceByName(traceName);
            if (trace != null) {
                call.reject(ERROR_TRACE_NAME_ALREADY_ASSIGNED);
                return;
            }
            implementation.startTrace(traceName);
            call.resolve();
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void stopTrace(PluginCall call) {
        try {
            String traceName = call.getString("traceName");
            if (traceName == null) {
                call.reject(ERROR_TRACE_NAME_MISSING);
                return;
            }
            Trace trace = implementation.getTraceByName(traceName);
            if (trace == null) {
                call.reject(ERROR_TRACE_NOT_FOUND);
                return;
            }
            implementation.stopTrace(traceName);
            call.resolve();
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void incrementMetric(PluginCall call) {
        try {
            String traceName = call.getString("traceName");
            String metricName = call.getString("metricName");
            Integer incrementBy = call.getInt("incrementBy", 1);
            if (traceName == null) {
                call.reject(ERROR_TRACE_NAME_MISSING);
                return;
            }
            if (metricName == null) {
                call.reject(ERROR_METRIC_NAME_MISSING);
                return;
            }
            Trace trace = implementation.getTraceByName(traceName);
            if (trace == null) {
                call.reject(ERROR_TRACE_NOT_FOUND);
                return;
            }
            implementation.incrementMetric(traceName, metricName, incrementBy);
            call.resolve();
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void setEnabled(PluginCall call) {
        try {
            Boolean enabled = call.getBoolean("enabled");
            if (enabled == null) {
                call.reject(ERROR_ENABLED_MISSING);
                return;
            }
            implementation.setEnabled(enabled);
            call.resolve();
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void isEnabled(PluginCall call) {
        try {
            Boolean enabled = implementation.isEnabled();
            JSObject result = new JSObject();
            result.put("enabled", enabled);
            call.resolve(result);
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void putAttribute(PluginCall call) {
        try {
            String traceName = call.getString("traceName");
            if (traceName == null) {
                call.reject(FirebasePerformancePlugin.ERROR_TRACE_NAME_MISSING);
                return;
            }
            String attribute = call.getString("attribute");
            if (attribute == null) {
                call.reject(FirebasePerformancePlugin.ERROR_ATTRIBUTE_MISSING);
                return;
            }
            String value = call.getString("value");
            if (value == null) {
                call.reject(FirebasePerformancePlugin.ERROR_VALUE_MISSING);
                return;
            }
            Trace trace = implementation.getTraceByName(traceName);
            if (trace == null) {
                call.reject(ERROR_TRACE_NOT_FOUND);
                return;
            }
            FirebasePerformance.putAttribute(trace, attribute, value);
            call.resolve();
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void getAttribute(PluginCall call) {
        try {
            String traceName = call.getString("traceName");
            if (traceName == null) {
                call.reject(FirebasePerformancePlugin.ERROR_TRACE_NAME_MISSING);
                return;
            }
            String attribute = call.getString("attribute");
            if (attribute == null) {
                call.reject(FirebasePerformancePlugin.ERROR_ATTRIBUTE_MISSING);
                return;
            }
            Trace trace = implementation.getTraceByName(traceName);
            if (trace == null) {
                call.reject(ERROR_TRACE_NOT_FOUND);
                return;
            }
            String value = FirebasePerformance.getAttribute(trace, attribute);
            JSObject result = new JSObject();
            result.put("value", value);
            call.resolve(result);
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void getAttributes(PluginCall call) {
        try {
            String traceName = call.getString("traceName");
            if (traceName == null) {
                call.reject(ERROR_TRACE_NAME_MISSING);
                return;
            }
            Trace trace = implementation.getTraceByName(traceName);
            if (trace == null) {
                call.reject(ERROR_TRACE_NOT_FOUND);
                return;
            }
            Map<String, String> attributesMap = FirebasePerformance.getAttributes(trace);
            JSObject result = new JSObject();
            JSObject resultMap = new JSObject();
            for (String attribute : attributesMap.keySet()) {
                resultMap.put(attribute, attributesMap.get(attribute));
            }
            result.put("result", resultMap);
            call.resolve(result);
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void removeAttribute(PluginCall call) {
        try {
            String traceName = call.getString("traceName");
            if (traceName == null) {
                call.reject(FirebasePerformancePlugin.ERROR_TRACE_NAME_MISSING);
                return;
            }
            String attribute = call.getString("attribute");
            if (attribute == null) {
                call.reject(FirebasePerformancePlugin.ERROR_ATTRIBUTE_MISSING);
                return;
            }
            Trace trace = implementation.getTraceByName(traceName);
            if (trace == null) {
                call.reject(ERROR_TRACE_NOT_FOUND);
                return;
            }
            FirebasePerformance.removeAttribute(trace, attribute);
            call.resolve();
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void putMetric(PluginCall call) {
        try {
            String traceName = call.getString("traceName");
            if (traceName == null) {
                call.reject(FirebasePerformancePlugin.ERROR_TRACE_NAME_MISSING);
                return;
            }
            String metricName = call.getString("metricName");
            if (metricName == null) {
                call.reject(FirebasePerformancePlugin.ERROR_METRIC_NAME_MISSING);
                return;
            }
            Double num = call.getDouble("num");
            if (num == null) {
                call.reject(FirebasePerformancePlugin.ERROR_NUM_MISSING);
                return;
            }
            Trace trace = implementation.getTraceByName(traceName);
            if (trace == null) {
                call.reject(ERROR_TRACE_NOT_FOUND);
                return;
            }
            FirebasePerformance.putMetric(trace, metricName, (long) Math.floor(num));
            call.resolve();
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void getMetric(PluginCall call) {
        try {
            String traceName = call.getString("traceName");
            if (traceName == null) {
                call.reject(FirebasePerformancePlugin.ERROR_TRACE_NAME_MISSING);
                return;
            }
            String metricName = call.getString("metricName");
            if (metricName == null) {
                call.reject(FirebasePerformancePlugin.ERROR_METRIC_NAME_MISSING);
                return;
            }
            Trace trace = implementation.getTraceByName(traceName);
            if (trace == null) {
                call.reject(ERROR_TRACE_NOT_FOUND);
                return;
            }
            long value = FirebasePerformance.getMetric(trace, metricName);
            JSObject result = new JSObject();
            result.put("value", value);
            call.resolve(result);
        } catch (Exception exception) {
            Logger.error(TAG, exception.getMessage(), exception);
            call.reject(exception.getMessage());
        }
    }

    @PluginMethod
    public void record(PluginCall call) {
        call.unimplemented("Not implemented on Android.");
    }
}
